File: frontend/src/services/processService.ts
import apiClient from './api';

export interface Process {
  processId: number;
  processCode: string;
  processName: string;
  processType?: string;
  sequenceOrder: number;
  isActive: boolean;
  tenantId: string;
  tenantName: string;
  remarks?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ProcessCreateRequest {
  processCode: string;
  processName: string;
  processType?: string;
  sequenceOrder?: number;
  remarks?: string;
}

export interface ProcessUpdateRequest {
  processName?: string;
  processType?: string;
  sequenceOrder?: number;
  remarks?: string;
}

const processService = {
  // Get all processes (ordered by sequence)
  getProcesses: async (): Promise<Process[]> => {
    const response = await apiClient.get<Process[]>('/processes');
    return response.data;
  },

  // Get active processes only
  getActiveProcesses: async (): Promise<Process[]> => {
    const response = await apiClient.get<Process[]>('/processes/active');
    return response.data;
  },

  // Get process by ID
  getProcess: async (id: number): Promise<Process> => {
    const response = await apiClient.get<Process>(`/processes/${id}`);
    return response.data;
  },

  // Get process by code
  getProcessByCode: async (processCode: string): Promise<Process> => {
    const response = await apiClient.get<Process>(`/processes/code/${processCode}`);
    return response.data;
  },

  // Create new process
  createProcess: async (process: ProcessCreateRequest): Promise<Process> => {
    const response = await apiClient.post<Process>('/processes', process);
    return response.data;
  },

  // Update process
  updateProcess: async (id: number, process: ProcessUpdateRequest): Promise<Process> => {
    const response = await apiClient.put<Process>(`/processes/${id}`, process);
    return response.data;
  },

  // Delete process
  deleteProcess: async (id: number): Promise<void> => {
    await apiClient.delete<void>(`/processes/${id}`);
  },

  // Activate process
  activateProcess: async (id: number): Promise<Process> => {
    const response = await apiClient.post<Process>(`/processes/${id}/activate`);
    return response.data;
  },

  // Deactivate process
  deactivateProcess: async (id: number): Promise<Process> => {
    const response = await apiClient.post<Process>(`/processes/${id}/deactivate`);
    return response.data;
  },
};

export default processService;

File: backend/src/main/java/kr/co/softice/mes/domain/service/ProcessService.java
package kr.co.softice.mes.domain.service;

import kr.co.softice.mes.common.exception.BusinessException;
import kr.co.softice.mes.common.exception.ErrorCode;
import kr.co.softice.mes.domain.entity.ProcessEntity;
import kr.co.softice.mes.domain.repository.ProcessRepository;
import kr.co.softice.mes.domain.repository.TenantRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Process Service
 * 공정 마스터 서비스
 *
 * @author Moon Myung-seop
 */
@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProcessService {

    private final ProcessRepository processRepository;
    private final TenantRepository tenantRepository;

    /**
     * Find all processes by tenant ID
     */
    public List<ProcessEntity> findByTenant(String tenantId) {
        return processRepository.findByTenantIdWithTenantOrderBySequence(tenantId);
    }

    /**
     * Find active processes by tenant ID
     */
    public List<ProcessEntity> findActiveByTenant(String tenantId) {
        return processRepository.findByTenantIdAndIsActiveWithTenant(tenantId, true);
    }

    /**
     * Find process by ID
     */
    public Optional<ProcessEntity> findById(Long processId) {
        return processRepository.findById(processId);
    }

    /**
     * Find process by process code
     */
    public Optional<ProcessEntity> findByProcessCode(String tenantId, String processCode) {
        return processRepository.findByTenant_TenantIdAndProcessCode(tenantId, processCode);
    }

    /**
     * Create new process
     */
    @Transactional
    public ProcessEntity createProcess(ProcessEntity process) {
        log.info("Creating process: {} for tenant: {}",
            process.getProcessCode(), process.getTenant().getTenantId());

        // Check duplicate
        if (processRepository.existsByTenantAndProcessCode(process.getTenant(), process.getProcessCode())) {
            throw new BusinessException(ErrorCode.PROCESS_ALREADY_EXISTS);
        }

        return processRepository.save(process);
    }

    /**
     * Update process
     */
    @Transactional
    public ProcessEntity updateProcess(ProcessEntity process) {
        log.info("Updating process: {}", process.getProcessId());

        if (!processRepository.existsById(process.getProcessId())) {
            throw new BusinessException(ErrorCode.PROCESS_NOT_FOUND);
        }

        return processRepository.save(process);
    }

    /**
     * Delete process
     */
    @Transactional
    public void deleteProcess(Long processId) {
        log.info("Deleting process: {}", processId);
        processRepository.deleteById(processId);
    }

    /**
     * Activate process
     */
    @Transactional
    public ProcessEntity activateProcess(Long processId) {
        ProcessEntity process = processRepository.findById(processId)
            .orElseThrow(() -> new BusinessException(ErrorCode.PROCESS_NOT_FOUND));

        process.setIsActive(true);
        return processRepository.save(process);
    }

    /**
     * Deactivate process
     */
    @Transactional
    public ProcessEntity deactivateProcess(Long processId) {
        ProcessEntity process = processRepository.findById(processId)
            .orElseThrow(() -> new BusinessException(ErrorCode.PROCESS_NOT_FOUND));

        process.setIsActive(false);
        return processRepository.save(process);
    }
}
